domains/certificates/Query.API/API/DataSyncSyncer/Persistence/SyncState.cs

```diff
@@ -1,33 +1,46 @@
 using System;
-using System.Collections.Generic;
 using System.Linq;
-using API.DataSyncSyncer.Client.Dto;
+using System.Threading.Tasks;
+using API.MasterDataService;
+using API.Query.API.Projections;
 using Marten;
+using Microsoft.Extensions.Logging;
 
 namespace API.DataSyncSyncer.Persistence;
 
 public class SyncState : ISyncState
 {
-    private readonly Dictionary<string, DateTimeOffset> periodStartTimeDictionary = new();
+    private readonly IDocumentStore documentStore;
+    private readonly ILogger<SyncState> logger;
 
-    public void SetNextPeriodStartTime(List<DataSyncDto> measurements, string GSRN)
+    public SyncState(IDocumentStore documentStore, ILogger<SyncState> logger)
     {
-        if (measurements.IsEmpty())
-        {
-            return;
-        }
-
-        var newestMeasurement = measurements.Max(m => m.DateTo);
-        periodStartTimeDictionary[GSRN] = DateTimeOffset.FromUnixTimeSeconds(newestMeasurement);
+        this.documentStore = documentStore;
+        this.logger = logger;
     }
 
-    public long GetPeriodStartTime(string GSRN, DateTimeOffset meteringPointOnboardedStartDate)
+    public async Task<long?> GetPeriodStartTime(MasterData masterData)
     {
-        if (periodStartTimeDictionary.TryGetValue(GSRN, out var periodStart))
+        try
+        {
+            await using var querySession = documentStore.QuerySession();
+
+            var projection = await querySession.LoadAsync<CertificatesByOwnerView>(masterData.MeteringPointOwner);
+            if (projection == null)
+                return masterData.MeteringPointOnboardedStartDate.ToUnixTimeSeconds();
+
+            var maxDateTo = projection.Certificates.Values
+                .Where(c => masterData.GSRN.Equals(c.GSRN, StringComparison.InvariantCultureIgnoreCase))
+                .Select(c => c.DateTo)
+                .DefaultIfEmpty(0)
+                .Max();
+
+            return Math.Max(maxDateTo, masterData.MeteringPointOnboardedStartDate.ToUnixTimeSeconds());
+        }
+        catch (Exception e)
         {
-            return periodStart.ToUnixTimeSeconds();
+            logger.LogWarning("Failed reading from database. Exception: {exception}", e);
+            return null;
```

Comment:
Should this return null instead of raising an exception?

Reply:
This is running in a background service and an unhandled exception will stop the background service and the application. If we throw an exception here, then something outside needs to handle it. 

I think it is better to catch this at the lower levels. That is why it can return null.